sprint 2

Damir-prg opened the pull request:
Контекст переписан на использование Redux
- Redux-toolkit✅
- Redux-thunk✅

Добавлено использование DnD ✅

Добавлено API для создание заказа ✅

Damir-prg committed:
feat: add packages & path aliases

Damir-prg committed:
feat: add API endpoints

Damir-prg committed:
feat: add redux store and slices

Damir-prg committed:
feat: changed context to redux store

Damir-prg committed:
feat: add handle viewed ingredient

Damir-prg committed:
feat: add scroll handlers

Damir-prg committed:
feat: add ingredients sorting

Damir-prg committed:
feat: add order request

Damir-prg committed:
part save

Damir-prg committed:
feat: update README

gennady-bars commented on src/services/reducers/selectedIngredientsSlice.ts:
```diff
@@ -0,0 +1,84 @@
+      state: TSliceSelectedIngredients,
+      action: PayloadAction<TIngredient>
+    ) {
+      state.selectedIngredients.push({ ...action.payload, __key: uuidv4() });
```
Редьюсер должен быть **чистой функцией**. Генерировать `uuid` в самом редьюсере **нельзя**. Это делает редьюсер нечистой функцией. Нужно генерировать uuid при создании экшена (желательно, в **`action creator`**). И созданный uuid должен приходить в экшене в `payload` или как отдельное поле, или сразу как часть объекта ингредиента. Посоветуйтесь с наставником, пожалуйста, если не получится
```js

// вот это - `action creator`
export const addIngridient = (item) => {
    return {
        type: ADD_INGRIDIENT,
        payload: {
            ...item, // используем `spread`, чтобы поменять ссылку на объект. Таким образом `redux` обновит его в хранилище
           uniqueId: uuid4()  // и добавляем в объект новое поле, которое потом будет использовано в `key`
        }
    }
}


```

А если используете `createSlice`, то это делается в `prepare`

```ts

prepare: (ingredient: TIngredient) => {
        return { payload: { ...ingredient, uniqueId: uuid4() } };
      }


```

Подробнее про `prepare` тут https://redux-toolkit.js.org/api/createSlice#customizing-generated-action-creators

gennady-bars commented on src/components/BurgerConstructor/ui/Bun/Bun.tsx:
```diff
@@ -1,20 +1,29 @@
-import { ComponentProps, FC } from "react";
+import { useCallback } from "react";
+import { useDispatch, useSelector } from "react-redux";
```
**МОЖНО ЛУЧШЕ**

`import { useDispatch, useSelector } from 'react-redux';` это нужно убрать из проекта.  Хуки `useDispatch, useSelector` нужно типизировать **1 раз** в отдельном файле и заменить их во всем проекте на свои, а не из библиотеки `react-redux` .  Вот ссылка на документацию по их типизации https://redux.js.org/usage/usage-with-typescript#define-typed-hooks

gennady-bars commented on src/components/BurgerConstructor/ui/Bun/Bun.tsx:
```diff
@@ -1,20 +1,29 @@
+export const Bun: FC<TBunProps> = ({ orientation }) => {
+  const dispatch: AppDispatch = useDispatch();
+  const { selectedBun } = useSelector(
+    (state: RootState) => state.selectedIngredients
```
**МОЖНО ЛУЧШЕ**

`(state: RootState)` уже не нужно будет типизировать вручную стейт, когда замените везде хук `useSelector`. Нужно будет удалить дублирование типизации

gennady-bars commented on src/components/BurgerConstructor/ui/ConstructorTotal/ConstructorTotal.tsx:
```diff
@@ -1,18 +1,38 @@
 import classes from "./constructorTotal.module.css";
 
 export const ConstructorTotal: FC = () => {
   const [modalState, setModalState] = useState(false);
```
**Можно лучше**

Если интересно, можно создать **кастомный хук** для управления модальным окном:

```js
import { useState, useCallback } from "react";

// кастомные хуки всегда должны начинаться с глагола `use`, чтобы реакт понял, что это хук. Он следит за их вызовами
export const useModal = () => {
  const [isModalOpen, setIsModalOpen] = useState(false);

  // `useCallback` нужен для того, чтобы зафиксировать ссылку на функцию. Таким образом уменьшится кол-во перерисовок компонента, куда будет передана эта функция
  const openModal = useCallback(() => {
    setIsModalOpen(true);
  }, []);

  const closeModal = useCallback(() => {
    setIsModalOpen(false);
  }, []);

  return {
    isModalOpen,
    openModal,
    closeModal,
  };
};

```

Этот код помещают в отдельный файл `useModal.js` в папке `hooks` и импортируют функцию туда, где нужно создать показатели открытости/закрытости попапа (у каждого попапа они свои)

И теперь нужно просто вызывать хук в том компоненте, где нужно управлять попапом. Всего одна строчка кода будет (под каждый попап своя):

```js
  const { isModalOpen, openModal, closeModal } = useModal();

````

gennady-bars commented on src/components/BurgerConstructor/ui/Ingredients/Ingredients.tsx:
```diff
@@ -25,8 +40,12 @@ export const Ingredients: FC<TIngredientsProps> = ({ selectedIngredients }) => {
+      {selectedIngredients.map((item, index) => (
+        <Ingredient
+          selectedIngredient={item}
+          key={item.__key}
```
`          key={item.__key}
`

Отлично, что указываете **уникальный** `__key` для ингредиентов в заказе. Это гарантирует 100% логику `DnD`

gennady-bars commented on src/api/ingredients/index.ts:
```diff
@@ -13,4 +18,20 @@ export const ingredientsApi = {
+      },
+    });
+
+    if (response.ok) {
+      return response.json();
+    }
+
+    return { success: false };
+  },
```
```js
.then(res => {
                if (res.ok) {
                    return res.json();
                }
                return Promise.reject(`Ошибка ${res.status}`);
            })

```

У Вас повторяется код **проверки ответа** во всех запросах. Чтобы этого избежать, нужно сделать отдельную функцию `checkResponse`, которая будет проверять ответ от сервера:

```js
function checkResponse(res) {
        // тут проверка ответа
    }

```

Ее нужно поместить в папку `utils` и импортировать туда, где нужна проверка на `ok`.

и теперь можно удалить кучу ненужного повторного кода и заменить его этим:

```js
 .then(checkResponse)

```

Обратите внимание, что передается только **ссылка** на функцию. Не нужно ее вызывать. Она сама вызовется, так как в `then` нужно передавать именно **функцию**, а не **вызов** функции.

-----------------------------------------

**Можно лучше**

Можно сделать **универсальную** функцию запроса с **проверкой ответа**, чтобы не дублировать эту проверку в каждом запросе:

```js

  function request(url, options) {
    // принимает два аргумента: урл и объект опций, как и `fetch`
    return fetch(url, options).then(checkResponse)
  }

```

И теперь просто нужно заменить все `fetch` на `request` и убрать дублирование проверки на `ok`. Все остальное будет без изменений. Код станет чище

И даже можно поместить внутрь `baseUrl`, чтобы не дублировать его в каждом запросе. Тогда нужно будет передавать просто `endpoint` в вызов

gennady-bars commented on src/api/ingredients/index.ts:
```diff
@@ -13,4 +18,20 @@ export const ingredientsApi = {
     }
```
У Вас повторяется код **проверки ответа** во всех запросах. Чтобы этого избежать, нужно сделать отдельную функцию `checkResponse`, которая будет проверять ответ от сервера:

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Посмотрите на **гитхабе** все комментарии к коду (нужно прокрутить вниз страницу там)

Работа проделана огромная:



* Отлично, что указываете **уникальный** `__key` для ингредиентов в заказе. Это гарантирует 100% логику `DnD` 
* Хорошая логика редакса и `DnD`


но есть некоторые недочеты:




* В верхней булке нужно добавлять текст `верх`, а в нижней - `низ`. Посмотрите в фигме, пожалуйста
* Редьюсер должен быть **чистой функцией**. Генерировать `uuid` в самом редьюсере **нельзя**. Это делает редьюсер нечистой функцией. Нужно генерировать uuid при создании экшена (желательно, в **`action creator`**). И созданный uuid должен приходить в экшене в `payload` или как отдельное поле, или сразу как часть объекта ингредиента. Посоветуйтесь с наставником, пожалуйста, если не получится
* У Вас повторяется код **проверки ответа** во всех запросах. Чтобы этого избежать, нужно сделать отдельную функцию `checkResponse`, которая будет проверять ответ от сервера:





**МОЖНО ЛУЧШЕ**

* Желательно **очищать конструктор** после успешного получения номера заказа с сервера в блоке `then`, чтобы пользователь мог следующий заказ сделать, не удаляя старые ингредиенты
* лучше не давать нажать на кнопку `Оформить заказ`, пока пользователь не добавил булку в заказ.

Damir-prg committed:
refactor: Правки с ревью

Bun: Добавлен текст "верх" и "низ"
ingredientsApi: Добавлена функция checkResponse для проверки ответа в запросах
selectedIngredientsSlice: Добавлен prepare
ConstructorTotal: Добавлена очистка конструктора после оформления заказа
services/hooks:  Типизированы хуки useDispatch, useSelector

Damir-prg commented:
Bun: Добавлен текст "верх" и "низ"
ingredientsApi: Добавлена функция checkResponse для проверки ответа в запросах
selectedIngredientsSlice: Добавлен prepare
ConstructorTotal: Добавлена очистка конструктора после оформления заказа и disabled на кнопку если конструктор пустой
services/hooks:  Типизированы хуки useDispatch, useSelector

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.
